Add Nix Flake Support to Espanso

delafthi opened the pull request:
This pull request introduces a Nix flake for the Espanso project, providing a development environment and build capabilities with nix.

#### Key Features
-   **Development Environment**: Set up with `nix shell`, containing all dependencies for building Espanso. Automatic activation with `direnv`.
-   **Build Capability**: Espanso can be built using `nix build`. Note that the build is not cached, resulting in a full rebuild upon any changes.

#### Note
This flake is a work in progress and may require further refinement. I am submitting it as a draft PR to facilitate discussion around the potential adoption of Nix in the project.

#### References
-   The structure of this flake was inspired by the implementation found in [sxyazi/yazi](https://github.com/sxyazi/yazi).
-   The package definition located in `nix/espanso.nix` is adapted from the package declaration in [nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/es/espanso/package.nix).

delafthi committed:
:sparkles: add nix flake

- add a flake that only takes nixpkgs as input
- use `nixfmt-rfc-style` as nix formatter

AucaCoyan commented:
Thank you! I was looking forward adding nix to the repo, but I'm a beginner in the nix ecosystem😊.

I read everything and so far looks great. I have just one question: 
In espanso I want to follow the stable release of rust and clippy. If rust updates to a new version and espanso refuses to compile, it should be treated as a bug. Does this make sense with the reproducibility of nix? Is it idiomatic of nix?

AucaCoyan commented:
also chiming @n8henrie who is a friend of the snow flakes 😎. Tell us your thoughts!

delafthi commented:
Finally, got [nix-community/naersk](https://github.com/nix-community/naersk) working, which should enable cached builds.
However, building with naersk breaks patching and running the build scripts to create the macOS app bundle. I need to find a way to fix that.

delafthi commented:
I'm curious whether you only format Rust files, as I only see a `rustfmt.toml` file and `cargo fmt` in your CI workflow. Currently, [numtide/treefmt-nix](https://github.com/numtide/treefmt-nix), only runs `alejandra` (for formatting Nix files) and `rustfmt`.

I would like to propose adding additional formatters for better code consistency:

-  **C/C++**: Use `clang-format` which requires a `.clang-format` file. It's possible that the initial sources were formatted with a version not currently in the repo, which would be nice to get. Else we would need to agree on a style.
-  **Markdown/JSON/YAML**: Use `prettier` which requires a `.prettierrc` and `.prettierignore` files.
-  **Bash**: Use `shfmt` which gets style from `.editorconfig`.

Additionally, I highly recommend adding a `.editorconfig` file, as it is widely supported by various editors and defines general formatting conventions.

n8henrie commented:
Oh man, this is really exciting! Below is my (opinionated) take.

For context, I'm one of the maintainers of the [nixpkgs package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/es/espanso/package.nix), the [NixOS module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/desktops/espanso.nix), and the [home-manager module](https://github.com/nix-community/home-manager/blob/master/modules/services/espanso.nix) for Espanso. That said, I'm still a learner in the nix community, just like with rust -- none of this is canon.

My love for nix is that it keeps pieces from moving under my feet. For example I have been working on [this little embedded rust project](https://github.com/n8henrie/esp32c3-rust-mqtt) that runs on an ESP32C3. Thanks to `nix`, I can get distracted for a few months but then come back and pick the project back up, and if it built before it still builds, irrespective of changes to the host system (be that MacOS or Linux).

A few thoughts on this PR:

> Development Environment: Set up with nix shell, containing all dependencies for building Espanso. Automatic activation with direnv.

`nix develop` should be for the development environment (it is the equivalent of the old-style `nix-shell`). `nix shell` should provide a shell with the fully built `espanso` binary available (nix shell is kind of like running `nix build && export PATH=./result/bin:$PATH`). (EDIT: It looks like this feedback has already been incorporated into the PR.)

Personally, I dislike unnecessary dependencies; one of my favorite things about nix is that it does an *amazing* job avoiding breakages due to unanticipated changes in dependencies. Personally, while they are popular, I find little use for and hence avoid `flake-compat`, `flake-utils`, `treefmt-nix`, and the like. Most of the time the required functionality can be abstracted away in a simple pure-nix function that basically never breaks or changes beneath my feet. Here is my default template for a new rust project: https://github.com/n8henrie/flake-templates/blob/master/rust/flake.nix -- note the only input is `nixpkgs`. (note to self: update to include `nixfmt-rfc-style` as formatter).

I usually try to avoid naersk or any of the other build helpers for other languages. They have a habit of becoming unmaintained quickly. (I thought we were all supposed to be on dream2nix by now.)

The one dependency I *do* sometimes use for rust projects is oxalica, and I have a separate rust template for this: https://github.com/n8henrie/flake-templates/blob/master/rust-oxalica/flake.nix

Of note, I only use oxalica when I need a nightly compiler, which is otherwise tough to do with nix.

In my opinion, given the desire to stick with stable rust, I would advise *not* using oxalica. There is a *huge* nix-rust community, and the rustc packaged in nixpkgs/unstable is the stable compiler and is reliably kept up to date.

```
$ nix eval github:nixos/nixpkgs#rustc.version
"1.86.0"
```

- 1.86 was released on Apr 3: https://releases.rs/docs/1.86.0/
- PR to update nixpkgs to 1.86 on Apr 3, (before 0600 my time): https://github.com/NixOS/nixpkgs/pull/395734

I don't really understand the concerns about cached builds. IMO rust development should be done in a `nix develop` environment with the standard Rust tooling. I sometimes use `sccache` and link with `mold` (both avail in nixpkgs) to speed up build times on linux. Once development is finished, then one can `nix build` to make sure the project still compiles.

Just as an example of how simple things can be, I wrote this tiny script a few days ago and used it to compile Espanso on a super old weak machine (an Acer C720 Chromebook that I bought for $150 over a decade ago and converted to linux -- 
1.4 GHz, 2 cores, 2G RAM, 116G HDD):

```bash
#!/usr/bin/env nix-shell
#!nix-shell -i bash -p bash --pure
#!nix-shell -p cargo-make
#!nix-shell -p cargo
#!nix-shell -p rustc
#!nix-shell -p mold
#!nix-shell -p pkg-config
#!nix-shell -p dbus
#!nix-shell -p libxkbcommon
#!nix-shell -p wxGTK32
#!nix-shell -p rust-script

set -Eeuf -o pipefail

main() {
  cargo make \
    --env NO_X11=true \
    --profile release \
    -- \
    build-binary

  ./target/release/espanso daemon
}
main "$@"
```

So if that ancient anemic machine can work, I'm not sure the extra complexity for caching is always necessary.

So, I think this PR is really exciting, and I bet it's going to work well -- thank you @delafthi for your work here! If *I* were rewriting it for my own use, my version would be considerably more bare-bones -- likely with `nixpkgs` as the only flake input, with likely only 2 `.nix` files (`flake.nix` and `default.nix`). But then again, I'm somewhat allergic to complexity and extra dependencies, and I'm happy to have somewhat less nice of a dev experience in exchange for simplicity and reliability -- this approach is definitely not for everyone.

Nix takes a while to get used to, but once it's taken hold, it's hard to imagine going back. Hope this PR makes it into Espanso! Please let me know how I can help!

AucaCoyan commented:
### RE: delafthi and formatting

Yes, I like formatters, I would like to include them in every project! This is my first cpp project, so I'm learning everything (even for loops) while in the meantime I have 30k lines in C/C++/Obj-C I have to maintain 🤣 . You can imagine my surprise when I found there is (almost) no commonly agreed style guide for the C family like I'm used to in Python, Rust and even JavaScript with all it's controversy.

Nix and rust have no problem, they have a standard or well agreed style guide to format files, but with Cpp is not that easy. I made an attempt with clang-format in PR #1864 . It was a missclick merge, and I had to revert everything back.
Btw, just asking: how to do to format every file with the clang-format, without using `make`?

About javascript files, I prefer [biomejs](https://biomejs.dev/), with the same tool it comes with a linter and nice interfaces. Also I like to support rust projects ❤️ .

I'm onboard with .editorconfig! Bring that little guy to the party too.

With bash I've never used a formatter, so everything is fine with me. Whatever n8 and you agree upon

### RE: n8henrie and nix

> avoid flake-compat, flake-utils, treefmt-nix, and the like.

Yeah, I know the feeling. Ditching stuff is good, but I rather fail into the "community preferences". As I commented [in this thread](https://github.com/espanso/espanso/issues/2301#issuecomment-2816869052) I want to make different development tools (OS, editor, shell) compatible for anyone, if in my end it's just adding a `.config` folder with sensible defaults.

> But then again, I'm somewhat allergic to complexity and extra dependencies, and I'm happy to have somewhat less nice of a dev experience in exchange for simplicity and reliability -- this approach is definitely not for everyone.
>
> Nix takes a while to get used to, but once it's taken hold, it's hard to imagine going back. Hope this PR makes it into Espanso! Please let me know how I can help!

Great! Happy we can all meet in a common ground. I'll tell you, I have nixOS installed in my main pc, but I have to resolve some data synchronization issues with OneDrive before using as a daily driver. Can't wait to use it! 💯

AucaCoyan commented:
Whenever you are ready, mark the PR as ready for review and I'll merge it 🚀

delafthi commented:
# RE: @n8henrie
Thank you for your feedback. Here’s my take on things:

> For context, I'm one of the maintainers of the [nixpkgs package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/es/espanso/package.nix), the [NixOS module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/desktops/espanso.nix), and the [home-manager module](https://github.com/nix-community/home-manager/blob/master/modules/services/espanso.nix) for Espanso. That said, I'm still a learner in the nix community, just like with rust -- none of this is canon.

That's great to hear! I might have a few questions about the [package definition](https://github.com/NixOS/nixpkgs/blob/nixos-24.11/pkgs/by-name/es/espanso/package.nix#L144) in `nixpkgs`, but I’ll need to do some testing first.  
I’m also planning to add Espanso plugins to both `nixpkgs` and `home-manager` soon, but I'll open a separate issue to discuss that.

> `nix develop` should be for the development environment (it is the equivalent of the old-style `nix-shell`). `nix shell` should provide a shell with the fully built `espanso` binary available (nix shell is kind of like running `nix build && export PATH=./result/bin:$PATH`).

I would recommend against this approach for a couple of reasons:  
- As far as I understand, `nix develop` is meant for developing or debugging nix packages — more like an interactive `nix build`, where you can manually run individual build phases and investigate issues. I had to do some digging myself and found [this great article](https://blog.ysndr.de/posts/guides/2021-12-01-nix-shells/) that explains it well.  
- The default `nix shell` should ideally be as lightweight as possible to minimize the time it takes to enter the shell. It should only include what's needed to develop (tooling) and build (dependencies) the project. Automatically building Espanso as part of the shell setup would significantly slow this down. That said, having a separate shell that includes the built Espanso binary sounds like a good idea — I'd recommend providing an alternative shell specifically for testing.

> (EDIT: It looks like this feedback has already been incorporated into the PR.)

Currently, Espanso itself isn't built or exported into the default shell — only its build dependencies are included.

> Personally, I dislike unnecessary dependencies; one of my favorite things about nix is that it does an *amazing* job avoiding breakages due to unanticipated changes in dependencies. While they are popular, I find little use for and hence avoid `flake-compat`, `flake-utils`, `treefmt-nix`, and the like. Much of the time the required functionality can be abstracted away in a simple pure-nix function that basically never breaks or changes beneath my feet. Here is my default template for a new rust project: https://github.com/n8henrie/flake-templates/blob/master/rust/flake.nix -- note the only input is `nixpkgs`. (note to self: update to include `nixfmt-rfc-style` as formatter).

I'm fine with that approach. Personally, I tend to use some of Numtide’s packages to reduce boilerplate and simplify the flake setup.  
One exception is `flake-compat` — I find it useful for supporting non-flake systems without having to maintain separate `default.nix` and `shell.nix` files. How would you suggest keeping non-flake compatibility without duplicating definitions?

> I usually try to avoid naersk or any of the other build helpers for other languages. They have a habit of becoming unmaintained quickly. (I thought we were all supposed to be on dream2nix by now.)
> 
> The one dependency I *do* sometimes use for rust projects is oxalica, and I have a separate rust template for this: https://github.com/n8henrie/flake-templates/blob/master/rust-oxalica/flake.nix
> 
> Of note, I only use oxalica when I need a nightly compiler, which is otherwise tough to do with nix.
> 
> In my opinion, given the desire to stick with stable rust, I would advise *not* using oxalica. There is a *huge* nix-rust community, and the rustc packaged in nixpkgs/unstable is the stable compiler and is reliably kept up to date.

Sounds good to me. My thought was simply that, since we already have a `rust-toolchain.toml` file, it might make sense to incorporate it. But I’m fine either way.

delafthi commented:
### RE: @AucaCoyan

Thank you for your answer.  
I think it would be better to move this discussion to a separate [issue](#2314) to keep things organized.  
I originally brought up formatters here because I've been using `treefmt` so far.  
However, since we aim to minimize dependencies in our flake, formatting should probably be treated as a separate topic — except for formatting Nix files with `nix fmt`, which makes sense to keep in our flake as mentioned by @n8henrie.

AucaCoyan commented:
Yup! we both like separating concerns 😁

n8henrie commented:
Hi! Glad to have some dialog.

> As far as I understand, nix develop is meant for developing or debugging nix packages
...
> The default nix shell should ideally be as lightweight as possible to minimize the time it takes to enter the shell. It should only include what's needed to develop (tooling) and build (dependencies) the project.

I think you are misunderstanding here. The official nix manual makes this clear:

https://nix.dev/manual/nix/2.28/command-ref/new-cli/nix3-develop#description

> nix develop starts a bash shell that provides an interactive build environment nearly identical to what Nix would use to build [installable](https://nix.dev/manual/nix/2.28/command-ref/new-cli/nix#installables). Inside this shell, environment variables and shell functions are set up so that you can interactively and incrementally build your package.

It is indeed for developing and debugging nix packages -- like Espanso, as packaged by this flake. It should provide the tooling need to develop the package, in this case tools like rustc, cargo, the dependencies like wx, libdus, and optionally convenience tooling like rust-analyzer, etc.

This is analogous to `nix-shell`, which specifies at https://nix.dev/manual/nix/2.28/command-ref/nix-shell#disambiguation

> This man page describes the command nix-shell, which is distinct from nix shell. For documentation on the latter, run nix shell --help or see man nix3-shell.

This is an older version of the manual (not sure why `nix shell` isn't in the manual for 2.28), but it spells it out pretty clearly:

> nix shell - run a shell in which the specified packages are available

https://nix.dev/manual/nix/2.13/command-ref/new-cli/nix3-shell

So when you run `nix shell github:nixos/nixpgks#espanso`, you get a shell session where `espanso` -- as a built binary -- is available. When you run `nix develop my-flake#espanso`, you should get an environment in which you have all necessary build tooling to build and develop espanso.

> How would you suggest keeping non-flake compatibility without duplicating definitions?

I don't usually maintain non-flake compatibility. In spite of their "experimental" status (now stabilized in Determinate nix), I that without flakes, nix's advantages are far less obvious given the mutable nature of channels. I do usually keep a separate `flake.nix` and ` default.nix`, the latter of which uses the `callPackage` pattern, which should make building from a non-flake system trivial.

delafthi commented:
You're absolutely right. What I meant was actually the environment you enter when using `direnv`, which evaluates `outputs.devShells.${system}.default`.  
I mistakenly assumed that `nix shell` was being executed, but it's actually `nix develop`. Sorry for the mix-up and any confusion 🫣.  
So yes, you're correct — `nix shell` drops you into a shell with Espanso installed. I edited my previous comment to prevent confusion and the spread of misinformation.

I'll make the changes tomorrow and ping you to review them.

delafthi commented:
Hey @n8henrie,

I just pushed an updated version of the flake. I tested the build on both macOS and Linux.

A few notes:

- In `nativeBuildInputs`, I removed `extra-cmake-modules` and `makeWrapper`. I'm not sure why `extra-cmake-modules` was needed — it seems to build fine without it.
- Regarding `makeWrapper`, I saw it's used in your `postInstall` hook on Linux:  
  [nixpkgs espanso package.nix L108–L130](https://github.com/NixOS/nixpkgs/blob/b024ced1aac25639f8ca8fdfc2f8c4fbd66c48ef/pkgs/by-name/es/espanso/package.nix#L108-L130)  
  Is it really necessary to wrap the binary and manually add those packages to the `PATH`? Since they’re already declared as runtime dependencies in `buildInputs`, they should be available without additional wrapping. In my testing, espanso works fine without the hook.
- I also removed `libnotify` from `buildInputs` — espanso still compiles successfully, and notifications appear to work as expected, at least on Gnome.

Let me know what you think!

delafthi commented:
I squashed all changes into a single commit and made a few additional improvements:

- Formatted all Nix files with `nixfmt-rfc-style`.
- Added `cargo-make` and `rust-script` to the dev shell (required for `cargo-make`).  
  
Let me know if I missed any other tools needed for development.

n8henrie commented:
@delafthi -- wow! Personally I think that's a lot better.

Stylistically, for the sake of consistency I usually collapse attrsets with only one attr:

```nix
inputs = {
    nixpkgs.url = ....
}
```
⬇️
```nix
inputs.nixpkgs.url = 
```

(this is as opposed to expanding `inputs = { nixpkgs = { url = `...)

This is just personal preference.

When there are changes, it certainly it makes for much easier diffs to do something like

```nix
inputs.nixpkgs.url = 
inputs.foobar.url = 
inputs.baz.url = 
```

as opposed to all the whitespace changes when you modify a level of indentation.

I'm fine either way, just style.

The other point I would consider, if you haven't looked at [my flake templates](https://github.com/n8henrie/flake-templates/blob/master/rust/flake.nix) yet, is that the `eachSystem` function I use give the same result as what you've done with `genAttrs`, but it spreads the `system` over all of the outputs. The benefit is that you get a significant reduction in repetitive `nixpkgsFor` calls everywhere, at the cost of a slightly less readable function (IMO).

Again, this is just style.

WRT your other points / question:

> In nativeBuildInputs, I removed extra-cmake-modules and makeWrapper. I'm not sure why extra-cmake-modules was needed — it seems to build fine without it.

We should make sure to test on the cross-product of aarch64, x86, linux, darwin, wayland, and x11 systems. If all of these build and run, then we should adopt these changes and upstream them to nixpkgs.

Speaking of which, I would ask -- please -- that we try to keep the nix derivation *here* as close to that used in nixpkgs as possible; this way we can upstream changes (and benefit from PRs there) with minimal friction. I think that would be an "everybody wins" situation.

> Regarding makeWrapper, I saw it's used in your postInstall hook on Linux: Is it really necessary to wrap the binary and manually add those packages to the PATH? Since they’re already declared as runtime dependencies in buildInputs, they should be available without additional wrapping. In my testing, espanso works fine without the hook.

Unsure -- this was added before I'd ever installed nix :) You'll have to follow the `blame`. Looks like maybe added in https://github.com/NixOS/nixpkgs/pull/126874/files

> I also removed libnotify from buildInputs — espanso still compiles successfully, and notifications appear to work as expected, at least on Gnome.

As per above -- let's test. *super* easy to ensure that it builds everywhere, harder to make sure that espanso works as intended.

Also, @AucaCoyan -- I think we should keep in mind a possible future in which nix is our build process for production binaries for linux and macos systems. I'm chugging away at a big CI PR, but there are *lots* of moving parts in there that nix could dramatically simplify.

n8henrie commented:
Also, I'm sure you are all aware, but I *love* how easy nix makes it to test PRs like this. You don't have to even have anything cloned anywhere.

```console
$ nix run github:delafthi/espanso/push-spmxwvxpnumu
espanso 2.2.3
Federico Terzi
A Privacy-first, Cross-platform Text Expander

USAGE:
    espanso [OPTIONS] [SUBCOMMAND]
...
```

O

n8henrie commented:
> I’ve also been thinking about better ways to test Wayland/X11 setups.

It might be worth investigating https://nixos.org/manual/nixos/unstable/#sec-call-nixos-test-in-nixos

It's been a while, but I contributed to the agenix tests and was really impressed with how much is possible: https://github.com/ryantm/agenix/blob/e600439ec4c273cf11e06fe4d9d906fb98fa097c/test/integration.nix#L70

It is entirely possible that we could have automatedend-to-end testing for at least some functionality; I'm not sure if `send_chars` is implemented in such a way that the input would be "expanded" or not.

Also easy to expose these as derivations for `nix flake check`.

AucaCoyan commented:
```
$ nix run github:delafthi/espanso/push-spmxwvxpnumu
```
Mind = blown. I was used to do `gh pr checkout 2311` (and I have autocompletion in my shell). 
That nix trick it's a really good unknown feature for me!

> a possible future in which nix is our build process for production binaries for linux and macos systems

Yes, a better build system is a quality of life upgrade everywhere!. I also browsed the possibility to use zig build system, though I didn't understood the benefits well enough for rust projects.

And about the GUI tests, it's been talked before, but little or not at all experimented. I vaguely remember somebody saying that is possible with Dev-containers, which are runnable in windows! (big win in that case).
Lastly, I like the mantra: testing is good, but impossible is better. If we by any means achieve that the software is un-buildable if you broke some promise, _that_ it would be the ideal scenario. Testing everything is also great, but it also requires significant amount of effort. I did a coverage report in #2287 and it was in the ~15% percent, so, we can do better!. This entire paragraph is enough to do another issue and lots of discussion

n8henrie commented:
> Mind = blown. I was used to do gh pr checkout 2311 (and I have autocompletion in my shell). That nix trick it's a really good unknown feature for me!

Oh yeah, it's awesome. In the GitHub web UI for a given PR, it even gives you a quick button to copy the user and branch to your clipboard:

<img width="556" alt="Screenshot 2025-04-21 at 13 50 15" src="https://github.com/user-attachments/assets/12b5810c-79f8-45ed-b5f5-5f62bae426ec" />

As long as their *fork* has a known name (in this case `espanso`), it makes it super easy to turn `delafthi:push-spmxwvxpnumu` into `github:delafthi/espanso/push-spmxwvxpnumu`.

> Yes, a better build system is a quality of life upgrade everywhere!. I also browsed the possibility to use zig build system, though I didn't understood the benefits well enough for rust projects.

The big thing with nix (particularly with flakes) is that your build dependencies don't move underneath your feet -- or at least not in a way that you can't easily roll back. 

> And about the GUI tests, it's been talked before, but little or not at all experimented.

I was meaning more of just integration tests -- making sure that `:espanso` gives you `Hello, espanso!` or what have you.

AucaCoyan commented:
> I was meaning more of just integration tests -- making sure that :espanso gives you Hello, espanso! or what have you.

Yes, but it's very difficult to prove the engine works without some minimal desktop environment, (as bare bones as it can get). You can test the functions work in rust code, or the interactions in the IPC, but everything combined is not easy. That's why I think sometimes it needs spining a vm. For example to test the clipboard setting, to give you an example. It's an area that needs love, there are lots of solutions out there, test-containers are also amazing and "doable"

n8henrie commented:
@AucaCoyan -- we're in agreement! You should read about the nixos tests and how they work, which is more or less exactly what you're describing! That's why they are the epitome of an *integration* test -- because done correctly, they test everything from package compilation (including running the package's unit tests) to actual functionality of the program in a VM.

I'm still not certain that it would work for Espanso (it might), but they are pretty cool, and worth knowing about.

AucaCoyan commented:
Are we ready to merge this? What's left if we are not yet?

n8henrie commented:
> We should make sure to test on the cross-product of aarch64, x86, linux, darwin, wayland, and x11 systems. If all of these build and run, then we should adopt these changes and upstream them to nixpkgs.

We could always merge with the understanding that this was provisional and untested on several platforms. (Not like it will have Windows support anyway!)

AucaCoyan commented:
Yeah, I'm fine with that. Dev containers are also "provisional" until somebody make use of it and configure them properly.
Merging!

AucaCoyan commented:
Yahoooo! we got nix! 🎉
